controller/eks-cluster-config-handler.go

```diff
@@ -1020,23 +1064,51 @@ func (h *Handler) updateUpstreamClusterState(upstreamSpec *eksv1.EKSClusterConfi
 		// happen together
 
 		ng := ngs[aws.StringValue(upstreamNg.NodegroupName)]
-		if ng.Version != nil {
-			if aws.StringValue(upstreamNg.Version) != desiredNgVersions[aws.StringValue(ng.NodegroupName)] {
-				attemptUpdatingNodegroups = true
-				_, err := eksService.UpdateNodegroupVersion(
-					&eks.UpdateNodegroupVersionInput{
-						NodegroupName: aws.String(aws.StringValue(ng.NodegroupName)),
-						ClusterName:   aws.String(config.Spec.DisplayName),
-						Version:       aws.String(desiredNgVersions[aws.StringValue(ng.NodegroupName)]),
-					},
-				)
+		ngVersionInput := &eks.UpdateNodegroupVersionInput{
+			NodegroupName: aws.String(aws.StringValue(ng.NodegroupName)),
+			ClusterName:   aws.String(config.Spec.DisplayName),
+		}
+
+		if upstreamNg.LaunchTemplate != nil {
+			upstreamTemplateVersion := aws.Int64Value(upstreamNg.LaunchTemplate.Version)
+			var err error
+			lt := ng.LaunchTemplate
+
+			if lt == nil && config.Status.ManagedLaunchTemplateID == aws.StringValue(upstreamNg.LaunchTemplate.ID) {
+				// In this case, Rancher is managing the launch template, so we check to see if we need a new version.
+				lt, err = newLaunchTemplateVersionIfNeeded(config, upstreamNg, ng, ec2Service)
 				if err != nil {
 					return config, err
 				}
-				continue
+			}
+
+			if lt != nil && (aws.StringValue(lt.ID) != aws.StringValue(upstreamNg.LaunchTemplate.ID) || aws.Int64Value(lt.Version) != upstreamTemplateVersion) {
+				attemptUpdatingNodegroups = true
+				ngVersionInput.LaunchTemplate = &eks.LaunchTemplateSpecification{
+					Id:      lt.ID,
+					Version: aws.String(strconv.FormatInt(*lt.Version, 10)),
+				}
+				if upstreamTemplateVersion > 0 {
+					templateVersionsToDelete = append(templateVersionsToDelete, aws.String(strconv.FormatInt(upstreamTemplateVersion, 10)))
+				}
+			}
+		}
+
+		if ng.Version != nil {
+			if aws.StringValue(upstreamNg.Version) != desiredNgVersions[aws.StringValue(ng.NodegroupName)] {
+				attemptUpdatingNodegroups = true
+				ngVersionInput.Version = aws.String(desiredNgVersions[aws.StringValue(ng.NodegroupName)])
 			}
 		}
 
+		if ngVersionInput.Version != nil || ngVersionInput.LaunchTemplate != nil {
```

Comment:
Have you tested if upgrading both at once works?

Reply:
nit: could move `attemptUpdatingNodegroup = true` here and remove from both blocks

Reply:
Yes, updating the launch template and version simultaneously works. Also move the `attemptUpdatingNodegroup = true` as you suggested.